### PR TITLE
yw-gnn-fmri-graph-neural-network

### PR DESCRIPTION
**Pull request by c-yuwei**
Implement graph neural network (GNN)

- [x] Create graph classifier using GNN with fMRI data

### CODE REVIEW

File: braph2genesis/neuralnetworks/_ClassifierGNN.gen.m [hunk 1]
```diff
@@ -0,0 +1,82 @@
+%% ¡header!
+ClassifierGNN < BaseNN (gnn, classifier with graph nerual network) is a classifier consists of graph neural network.
```
Comment (egolol) [inline braph2genesis/neuralnetworks/_ClassifierGNN.gen.m:2]: neural*

File: braph2genesis/neuralnetworks/_ClassifierGNN.gen.m [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+%%% ¡prop!
+TRAINING_ACCURACY (result, scalar) is the accuracy of the classifier obtained from the training data.
+
+%%% ¡prop!
+TEST_ACCURACY (result, scalar) is the accuracy of the classifier obtained from test data.
+
+%% ¡methods!
+function [score, prediction] = accuracy(gnn, YPred, target, classes)
+    % Decode probability vectors into class labels
+    prediction = onehotdecode(YPred, classes, 2);
```
Comment (egolol) [inline braph2genesis/neuralnetworks/_ClassifierGNN.gen.m:19]: functions should be easy to read. for example oneHotDecode or one_hot_decode, we use the latter.

File: braph2genesis/neuralnetworks/_ClassifierGNN.gen.m [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+function weights = initialize_glorot(gnn, sz,numOut,numIn,className)
+    arguments
+        gnn
+        sz
+        numOut
+        numIn
+        className = 'single'
+    end
+
```
Comment (egolol) [inline braph2genesis/neuralnetworks/_ClassifierGNN.gen.m:31]: this is what we usually do.
if nargin < 5 
className = 'single'
end

File: braph2genesis/neuralnetworks/_RegressorNN.gen.m [hunk 4]
```diff
@@ -0,0 +1,29 @@
+%% ¡header!
+RegressorNN < BaseNN (nn, regressor with nerual network) is a regressor consist of artificial neural network.
```
Comment (egolol) [inline braph2genesis/neuralnetworks/_RegressorNN.gen.m:2]: neural

File: braph2genesis/neuralnetworks/_ClassifierNN.gen.m [hunk 5]
```diff
@@ -0,0 +1,65 @@
+%% ¡header!
+ClassifierNN < BaseNN (nn, classifier with nerual network) is a classifier consists of artificial neural network.
```
Comment (egolol) [inline braph2genesis/neuralnetworks/_ClassifierNN.gen.m:2]: neural

File: braph2genesis/neuralnetworks/_ClassifierGNN.gen.m (hunk 2 above)
Comment (c-yuwei) [reply braph2genesis/neuralnetworks/_ClassifierGNN.gen.m:20]: Yes, agreed. The thing is that, this function 'onehotdecode' is a built-in function from matlab. Shall we remain this way or assemble it with 
function pred = one_hot_decode(args)
pred = onehotdecode(args)
end

File: braph2genesis/neuralnetworks/_RegressorNN.gen.m (hunk 4 above)
Comment (c-yuwei) [reply braph2genesis/neuralnetworks/_RegressorNN.gen.m]: done refer to 55aa78409182180ed9ba190199eb366a7384b5d7

File: braph2genesis/neuralnetworks/_ClassifierNN.gen.m (hunk 5 above)
Comment (c-yuwei) [reply braph2genesis/neuralnetworks/_ClassifierNN.gen.m]: done refer to 55aa78409182180ed9ba190199eb366a7384b5d7

**Comment by annacanal** [comment]
When I run the example_nn_classification_ST.m, at the end this error appears:

`Error using exportONNXNetwork (line 33)
exportONNXNetwork requires the Deep Learning Toolbox Converter for ONNX Model Format support package. To install this support package, use the Add-On Explorer.

Error in BaseNN/net_binary_transformer (line 707)
		    exportONNXNetwork(net,filename);

Error in ClassifierNN_ST/calculate_results (line 990)
		        neural_network = nn.net_binary_transformer(net);

Error in ClassifierNN_ST/calculateValue (line 911)
					value = calculate_results(nn);

Error in Element/get (line 863)
                        value = el.calculateValue(prop);

Error in Element/memorize (line 899)
            value = el.get(prop);

Error in example_nn_classification_ST (line 34)
nn_classifier_ST.memorize('NEURAL_NETWORK_ANALYSIS');`

So I understand I don't have the Deep learning toolbox. But when I did braph2genesis, I didn't receive any warning

File: braph2genesis/genesis/genesis.m [hunk 6]
```diff
@@ -166,6 +179,18 @@
...
+    if installedDLToolbox
+        nn_gen_list = getGenerators([source_dir fp 'src' fp 'nn']);
+        for i = 1:1:numel(nn_gen_list)
+            create_Element([source_dir fp 'src' fp 'nn' fp nn_gen_list{i}], [target_dir fp 'src' fp 'nn'])
+        end
+    else
+        warning('Deep Learning Toolbox is not installed. Please refer to <a href="matlab: web(''https://se.mathworks.com/products/deep-learning.html'') ">Deep Learning Toolbox</a> and <a href="matlab: web(''https://se.mathworks.com/matlabcentral/fileexchange/67296-deep-learning-toolbox-converter-for-onnx-model-format'') ">Deep Learning Toolbox Converter for ONNX Model Format</a>');
+    end
+    
     % graphs
```
Comment (annacanal) [inline braph2genesis/genesis/genesis.m:194]: I didn't have the toolbox installed and the warning didn't appear when I run braph2genesis. I am using a MAC

Comment (c-yuwei) [reply braph2genesis/genesis/genesis.m]: Thanks Anna, I have resolved this situation. Please refer to 712b0cc2a0adb707ecad233251b89f5f92621426.

**Comment by c-yuwei** [comment]
> > > > When I run the example_nn_classification_ST.m, at the end this error appears:
> > > > ```
> > > > Error using exportONNXNetwork (line 33)
> > > > exportONNXNetwork requires the Deep Learning Toolbox Converter for ONNX Model Format support package. To install this support package, use the Add-On Explorer.
> > > > 
> > > > Error in BaseNN/net_binary_transformer (line 707)
> > > > 		    exportONNXNetwork(net,filename);
> > > > 
> > > > Error in ClassifierNN_ST/calculate_results (line 990)
> > > > 		        neural_network = nn.net_binary_transformer(net);
> > > > 
> > > > Error in ClassifierNN_ST/calculateValue (line 911)
> > > > 					value = calculate_results(nn);
> > > > 
> > > > Error in Element/get (line 863)
> > > >                         value = el.calculateValue(prop);
> > > > 
> > > > Error in Element/memorize (line 899)
> > > >             value = el.get(prop);
> > > > 
> > > > Error in example_nn_classification_ST (line 34)
> > > > nn_classifier_ST.memorize('NEURAL_NETWORK_ANALYSIS');
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > So I understand I don't have the Deep learning toolbox. But when I did braph2genesis, I didn't receive any warning And I saw in the code of braph2genesis some lines checking if deep learning toolbox was installed and if not, throwing a warning
> > > 
> > > 
> > > Well it requires the toolbox: Deep Learning Toolbox Converter for ONNX Model Format; which works from R2018a and beyond (something also to note for BRAPH2 versions)
> > 
> > 
> > Yes, thanks anna, where should we note this information? Do you think I should add it in the warning message?
> 
> Now it appears! :) The last thing I would say about their warning is that it should be repeated at the end of braph2genesis, so the user does not miss it (I had to go up on the search because I was on purpose looking for it) ^^ or maybe it is not necessary, and a user that wants to use the neural networks will be aware of the dependency

Hi @annacanal thanks!
I changed the check mechanism of the deep learning toolboxes and now it shall work for both with and without the DL toolboxes.
Please refer to aa0344c3dcb2e2da39533fcd6722a7c32728a874.

**Pull request merged**